### ISSUE TITLE
Tracee fails on startup after signal other than sigint, doesn't have cleanup behavior

### ISSUE THREAD

**Issue by grantseltzer** [aquasecurity/tracee#447]
I ran `tracee --security-alerts`, hit ctrl-c, then ran it again and get the following error:

```
failed to add kprobe 'p:kprobes/psecurity_inode_unlink security_inode_unlink': -17
failed to create kprobe event: -17
```

I've tried running `make clean && make` but the problem still occurs. This is with latest in `main` branch  (`a87426a702aa1b69d38dbe1f96b8179f38471ea5`)

**Comment by grantseltzer** [comment]
Additional context:

```
[root@localhost]/home/vagrant/go/src/github.com/aquasecurity/tracee# uname -ar
Linux localhost.localdomain 5.8.15-301.fc33.x86_64 #1 SMP Thu Oct 15 16:58:06 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```

```
[root@localhost]/home/vagrant/go/src/github.com/aquasecurity/tracee# ls /sys/kernel/debug/tracing/events/kprobes/
enable  pcap_capable          psecurity_file_mprotect  psecurity_inode_unlink
filter  psecurity_bprm_check  psecurity_file_open      psecurity_mmap_addr
```
```
[root@localhost]/home/vagrant/go/src/github.com/aquasecurity/tracee# cat /sys/kernel/debug/tracing/kprobe_events
p:kprobes/psecurity_bprm_check security_bprm_check
p:kprobes/psecurity_file_open security_file_open
p:kprobes/pcap_capable cap_capable
p:kprobes/psecurity_mmap_addr security_mmap_addr
p:kprobes/psecurity_file_mprotect security_file_mprotect
p:kprobes/psecurity_inode_unlink security_inode_unlink
```

**Comment by grantseltzer** [comment]
Would we expect `/sys/kernel/debug/tracing/kprobe_events` to be removed when tracee is no longer running? Same with removal of `/sys/kernel/debug/tracing/events/kprobes/[event]`?

If so we can add cleanup on startup if tracee doesn't see another tracee process running

**Comment by yanivagman** [comment]
Using ctrl-c should have gone through Tracee's close function which calls t.bpfModule.Close(). This in turn deletes all the used kprobes. Failure to remove these created kprobes will result in the error you have seen.
It's exactly the same issue I discussed with @simar7, and we should find a way to avoid this from happening - either by global exception handling on exit or cleaning on start (If choosing the second option, we should give our probes unique names, so that we will not remove kprobes of other processes in the system)

**Comment by grantseltzer** [comment]
I will investigate/fix this

**Comment by grantseltzer** [comment]
Ah ok, what I didn't realize is that when I hit ctrl-c it was not on just `tracee --security-alerts` but I had piped it to `less`. So it has something to do with how signals are sent to programs exec'd by a shell command with pipes. I think we should have some cleanup process at the beginning regardless of how the cleanup step got skipped. 

With that in mind, I have a question about what you said: `"we should give our probes unique names, so that we will not remove kprobes of other processes in the system"`

Do we care to support having multiple invocations of Tracee? 

So at startup, we can check if there's another Tracee process running, and if not, clear all kprobes that start with `tracee_*` or something that we prefix them with.

**Comment by itaysk** [comment]
shell should send a SIGINT to both less and tracee if ctrl+c. but as a good practice, we could handle SIGPIPE as well.
cleanup before starting is a good idea

**Comment by yanivagman** [comment]
@grantseltzer there is one case which may require multiple instances of tracee to run and that is inspektor-gadget.
@ogozman has a working poc that integrates tracee into inspekto-gadget, and I think the only option for it to work is by having tracee to support multiple instances.
Also see the discussion at https://github.com/kinvolk/inspektor-gadget/issues/106 where we discussed exactly this.

**Comment by grantseltzer** [comment]
@itaysk Interestingly, quitting `less` sends sigsuspend followed by kill. And even more interestingly I determined that with `tracee --trace set=signals`! 

@yanivagman Got it! I will have a proposal/pr for this today or tomorrow.

**Comment by grantseltzer** [comment]
As it turns out this problem is fixed by using `AttachKprobe` instead of `AttachKprobeLegacy` in tracee.go. 

The reason is that `AttachKprobeLegacy` writes directly to the tracefs interface (`/sys/kernel/debug/tracing/kprobe_events`). The latter uses the libbpf function `bpf_program__attach_kprobe` which uses the perf_event_open interface for dynamically creating kprobes. In the latter there is no name collision and the cleanup happens automatically regardless of how the process exits. This also allows for multiple invocations of Tracee at the same time! 

This pushes the issue, can we up our minimal kernel version to 4.18 so we can get this added benefit?

**Comment by itaysk** [comment]
> Interestingly, quitting less sends sigsuspend followed by kill

sigsuspend and kill are just the syscalls involved in sending and handling the signal, it doesn't say *which* signal was sent to tracee, which (according to memory, might be inaccurate) should be either SIGINT or SIGPIPE depending on the order of termination.

**Comment by ogozman** [comment]
Hello, i think that it is not really a must to update minimal kernel to use perf probes, I have solved this issue with falling back perf from legacy if legacy fails like this
https://github.com/ogozman/tracee/blob/b72420a6d9f0cfc5896844adea361a497fb5f2c9/tracee/tracee.go#L831-L842

**Comment by grantseltzer** [comment]
@ogozman That's an interesting solution but it would still leave kprobes activated via tracefs with no consumer of them. I think the correct course of action would be to call the function that corresponds to the supported kernel. One scenario I can picture is an error occurs for a different reason when trying to install the probes via tracefs where calling `AttachKprobe` would not be the correct way of handling it.

As a consumer of tracee would updating minimal kernel be an issue for you? If so can you comment in #452 ?

**Comment by itaysk** [comment]
I've removed the bug label because tracee was not intended to support multiple instances until now.

**Comment by simar7** [comment]
> As it turns out this problem is fixed by using `AttachKprobe` instead of `AttachKprobeLegacy` in tracee.go.
> 
> The reason is that `AttachKprobeLegacy` writes directly to the tracefs interface (`/sys/kernel/debug/tracing/kprobe_events`). The latter uses the libbpf function `bpf_program__attach_kprobe` which uses the perf_event_open interface for dynamically creating kprobes. In the latter there is no name collision and the cleanup happens automatically regardless of how the process exits. This also allows for multiple invocations of Tracee at the same time!
> 
> This pushes the issue, can we up our minimal kernel version to 4.18 so we can get this added benefit?

My vote would be to push the minimal kernel version to 4.18 - if that's what it takes to help this move along. Currently if we run into this issue (not clean exit, multiple instances etc.) the UX is very poor because at least I don't know how to get into a sane state other than a reboot (high friction).

**Comment by grantseltzer** [comment]
@simar7 Agreed with what you said, but in the meantime you can fix this issue with `sudo bash -c "echo "" > /sys/kernel/debug/tracing/kprobe_events"`

**Comment by grantseltzer** [comment]
Closed by #551
